Real housing from improvements

bolbass opened the pull request:
#248 fixed finally! :smile:

SpaceOgre commented on Assets/UI/Panels/citypanel.lua:
```diff
@@ -1325,6 +1326,39 @@ function CQUI_UpdateSelectedCityCitizens( plotId:number )
+      end
+    end
+  end
+  CQUI_RealHousingFromImprovements = CQUI_TilesWithHousingCountX2 * 0.5;
```
It is good to avoid unnecessary global variables when possible, this method could return the value instead of setting the variable. This comment can be applied to all the new methods.

bolbass replied:
I tried to avoid it but didn't know how to do so. I'll check this out

SpaceOgre replied:
It is not that hard :), let me show you with an example:
`
function Main()
    var value = GetStuff(2);
    ... do something with value
end
function GetStuff(value:number)
    var value2x = value * 2;
    return value2x;
end
`

bolbass replied:
I've made all the changes and left the same function for all 4 files but I still can't reuse it from one file to all these 4 files. I've added it to `civ6common.lua` then tried to `Initialize` it in these files and use it's value. But it didnt work. Can you explain it?

SpaceOgre commented on Assets/UI/WorldView/citybannermanager.lua:
```diff
@@ -3244,6 +3245,44 @@ function OnCityBannerRightClick( playerID:number, cityID:number )
 
 -- ===========================================================================
+-- CQUI calculate real housing from improvements for each city
+function CQUI_AllCitiesRealHousingFromImprovements()
```
This function should take pCity as parameter since you only use the value for this one city anyway.

bolbass replied:
Got it

SpaceOgre commented on Integrations/URS/reportscreen.lua:
```diff
@@ -2130,6 +2131,45 @@ function OnToggleBonus()
 
+-- ===========================================================================
+-- CQUI calculate real housing from improvements for each city
+function CQUI_AllCitiesRealHousingFromImprovements()
```
Same as above, no need to calculate for all cities one you are only interested in the value for one.

bolbass commented:
I've made all the changes and left the same function for all 4 files but I still can't reuse it from one file to all these 4 files. I've added it to civ6common.lua then tried to Initialize it in these files and use it's value. But it didnt work. Can you explain how to do so? Fixed function is in this 79d16dda9cbae5c59378fed8ea1700c840fb990c commit

SpaceOgre commented:
I can try to. If you look at the top in `cqui_settingselement.lua`  you will see this row `include("Civ6Common")` this means that you can use functions from this file in the current one (might not be totally correct I'm not that good at Lua), like PopulateCheckBox and so on.

So try and add a function to Civ6Common and then call it from another file that have it included to start with. If you want to add a completely new file then it needs to be added in the modinfo file. 

Did it clear it up somewhat?

bolbass commented:
`Civ6Common` is already included to `city panel`, `overview`, `citybanner` and `reportscreen` files. I tried to add a function to `Civ6Common` and simply call it from these files but it didn't work.

SpaceOgre commented:
Did you try it with return value instead of global variable? The included file can't access the global variable from the calling file. 

Do you get any errors in the logs/firetuner when trying this?

bolbass commented:
The last attached file uses local variable:
`local CQUI_HousingFromImprovements = CQUI_RealHousingFromImprovements(selectedCity);`

bolbass commented:
@SpaceOgre it works now. The problem was I forgot to rename a variable in one row. Thanks! :+1:

SpaceOgre commented on Assets/UI/civ6common.lua:
```diff
@@ -1020,4 +1020,36 @@ function PopulateSlider(control, label, setting_name, data_converter, tooltip)
+  local CQUI_HousingFromImprovements = {};
+  local m_pCity = Players[Game.GetLocalPlayer()]:GetCities();
+  for i, pCity in m_pCity:Members() do
+
```
Not sure how Lua handles scopes but I think you might overwrite pCity that you pass by doing this and thus always return the last city. 

And since you already pass the city as parameter to the function this loop should not be needed no?

bolbass commented:
Something wrong with citybannermanager code changes. It works when start a game, on next turn, when load a saved game but when loading autosaves all citybanners miss.

SpaceOgre commented:
Can't help you anymore but point out that logs and firetuner are your friends in these situations. My computer with Civ on it is unplugged because of renovation and will be for a couple of days.

JHCD commented:
Sounds like a missing initialisation...
Could be, that there is a different behaviour between "load game" and "continious game".

bolbass commented:
Yes, it seems so because after disabling/enabling it on settings it works again

JHCD commented:
I remember something similar within another functionality...
But can't remember the topic...

SpaceOgre commented:
Yeah it might be that the citybanner is initialized before some of the things needed for the new functions and thus it crashes. As long as you can figure out what is causing the crash you should hopefully be able to create a check for it so init on load can run and after that is done it should work as intended. 

Could be tricky to track down the offending party though.
